### ISSUE TITLE
ssl_connect instead of ssl_disconnect?

### ISSUE THREAD

**Issue by DimitriPapadopoulos** [adrienverge/openfortivpn#239]
_ssl_disconnect_() is called once, at the beginning of _ssl_connect_():
https://github.com/adrienverge/openfortivpn/blob/61ac328efc3b7071f7de0a039f50a387d819a6ef/src/tunnel.c#L575-L577
I'm not certain about the logic of this. It looks like a hack to work around a missing _ssl_disconnect_() elsewhere.

At the same time _ssl_connect_() is called when terminating _run_tunnel_(), which doesn't make sense to me:
https://github.com/adrienverge/openfortivpn/blob/61ac328efc3b7071f7de0a039f50a387d819a6ef/src/tunnel.c#L790-L796
I believe this last call might be the reason for this Valgrind warning:
```
Open AF_INET socket 3: xxx.xxx.x.xx:45226 <-> xxx.xxx.xxx.xx:443
   at 0x5A0F567: socket (syscall-template.S:84)
   by 0x40A6B6: tcp_connect (tunnel.c:345)
   by 0x40A6B6: ssl_connect (tunnel.c:579)
   by 0x40B658: run_tunnel (tunnel.c:796)
   by 0x403F7A: main (main.c:428)
```

**Comment by DimitriPapadopoulos** [comment]
This is how I run __openfortivpn__ under Valgrind:
`sudo valgrind --track-fds=yes --track-origins=yes -v openfortivpn`

**Comment by mrbaseman** [comment]
I think for login and logout separate connections are made, independent of the vpn channel, with the goal to make the login/logout process more reliable, e.g. in cases when the tunnel breaks down for some reason

**Comment by mrbaseman** [comment]
the call to `ssl_disconnect` at the beginning of `ssl_connect` causes many disconnects and reconnects, especially since it is used in [http_request](https://github.com/adrienverge/openfortivpn/blob/master/src/http.c#L243) as well. 
Normally, this shouldn't be an issue as long as the tunnel is already up before `ssl_connect` is called. The calls in [run_tunnel](https://github.com/adrienverge/openfortivpn/blob/master/src/tunnel.c#L699) seem to be fine in my view. 
At the beginning, we can be sure that there is no tunnel there, so [ssl_disconnect](https://github.com/adrienverge/openfortivpn/blob/master/src/tunnel.c#L558) will just return because of the `if (!tunnel->ssl_handle) return;` statement. If the tunnel is up and `ssl_disconnect` is called, it also makes sure to clean up everything and sets this handle to `NULL`, so that's fine, too. 

I don't know exactly what happens if the underlying socket layer has a problem and is blocking, the tcp connection has been lost for some reason, etc... The return values and errors emitted by `SSL_shutdown` and `close` on the socket are not checked in `ssl_disconnect`. This might be a problem, especially at the end of the connection when the logout doesn't work, and the remote side believes that the client is still connected, and login sessions per user are limited. When the user hits the limit in this situation, he is not able to reconnect. However, we have just checked that in this situation all sessions of that user are terminated, so the next attempt to log in with correct credentials should work again.

Maybe there are more severe cases, e.g. when connecting over an unstable wifi connection. I could imagine that instabilities of the connection between "Step 0" and "Step 5" of the [run_tunnel](https://github.com/adrienverge/openfortivpn/blob/master/src/tunnel.c#L699) routine could turn the socket into a state in which _openfortivpn_ may not be able to continue. 

Anyhow, from a user perspective, when connecting from an insecure and unstable network I would expect any weird behavior until I see the message that the tunnel is finally up and running.

**Comment by DimitriPapadopoulos** [comment]
It would nice to re-run _openfortivpn_ under [Valgrind](http://valgrind.org/) one of these days and check whether it reports anything useful...

**Comment by mrbaseman** [comment]
In the meantime we have debugging output for the http requests with `-v -v`.  Maybe this helps somehow? 
I'm not sure anymore if my explanations [above](https://github.com/adrienverge/openfortivpn/issues/239#issuecomment-349660899l) are really correct. 
The ssl connects and disconnects are in different threads than the ppp tunnel, and they start first with the authentication.

**Comment by DimitriPapadopoulos** [comment]
See change caa21a6. Seems rather complex:
> We don't really understand Connection: close and thus don't find out
when the remote decided not to reuse the connection. Let's just try once
more if we encounter an EOF in a HTTP connection.
> 
> Also, always use a separate connection for the tunnel. The tunnel
connection is long-running and not really restartable so we can't really
just re-do it on EOF. Besides that, always close it when the tunnel
closes and don't use it for the /remote/logout request -- the server can't
tell that we're no longer doing PPP and are back to HTTP.

In short I understand  the server might disconnect on its own so we might have more calls to _ssl_connect_() than _ssl_diconnect_().
